### ISSUE TITLE
"Fork" should fire an exception when trying to consume inner iterables without consuming the container one

### ISSUE THREAD

**Issue by sithmel** [sithmel/iter-tools#198]
Using fork in a for .. of loop is highly wasteful because it forces to store all input iterable in memory.
```js
for (const f of fork(inputIterable)) {
  // whatever I do with f here will be cached in the buffer.
  // I can't consume the buffer because I am still able to create a new fork
  // and that will return the items from the beginning
}
```
The sensible way to use fork is using destructuring:
```js
const [f1, f2] =  fork(inputIterable)
```
For this reason I think it would be better to forbid to consume the forked iterables when the original iterable is still able to emit new "clones".
```js
for (const f of fork(inputIterable)) {
  f[Symbol.iterator]().next() // Error! You can't consume the forked iterable while you can still emit new iterables
}
```

**Comment by KSXGitHub** [comment]
This would surprise users, give unexpected bugs, limit use cases. I disagree.

**Comment by sithmel** [comment]
> give unexpected bugs

all the opposite. The purpose is to prevent the user using fork in a way that can unexpectedly suck up all the memory

> limit use cases (such as concurrency)

Why this should limit concurrency ?

The only usage that is forbidding is a self harmful one

**Comment by KSXGitHub** [comment]
> > give unexpected bugs
> 
> all the opposite. The purpose is to prevent the user using fork in a way that can unexpectedly suck up all the memory

I don't know how does the new implementation be memory efficient until I see it. But it would certainly surprise user when they get creative.

> Why this should limit concurrency?

Not between forks, but between forks and their container. Such case may seem rare to you, but we never know until it arises (i.e. when they get creative).

**Comment by sithmel** [comment]
The critical part is here:
https://github.com/sithmel/iter-tools/blob/master/src/fork.mjs#L62

fork is based on a queue object, where different consumers point to different position of the queue.
Whenever a new consumer is created, it starts with the beginning of the queue.

For this reason the entire queue is preserved in memory.

When finally the function noMoreConsumers https://github.com/sithmel/iter-tools/blob/master/src/fork.mjs#L62 is invoked, I am not forced to store the head of the queue and the garbage collector reclaims all unreachable queue items.

Before noMoreConsumers
```
queue.head = points to 1
consumer1 points to 3
consumer2 points to 7
queue:

1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7
```
after:
```
queue.head = null
consumer1 points to 3
consumer2 points to 7
queue:

3 -> 4 -> 5 -> 6 -> 7
```
1 and 2 are reclaimed by the garbage collector.

So in a nutshell, consuming the forked iterables when is still possible emitting new ones, means storing every single item in RAM. And this defeats the purpose of using this data structure.

**Comment by KSXGitHub** [comment]
Now that I notice, why didn't you call `exchange.noMoreConsumers` after you are sure that [every single fork is exhausted](https://github.com/sithmel/iter-tools/blob/3a9aa630667cfbebae35593642ed117528dbed6f/src/fork.mjs#L25).

And even if what I proposed isn't possible, memory is the last thing I concern as long as I don't use global variables.

**Comment by sithmel** [comment]
> why didn't you call `exchange.noMoreConsumers` after you are sure that [every single fork is exhausted](https://github.com/sithmel/iter-tools/blob/3a9aa630667cfbebae35593642ed117528dbed6f/src/fork.mjs#L25)?

Because it would like closing the gate after the chickens ran away :-)

> memory is the last thing I concern

If you are not concerned about memory, arrays are better from any point of view.

The only valid reason to use iterables is that we are dealing with **a lot** of data. Possibly more than you can fit in your process.

**Comment by conartist6** [comment]
I'm against this restriction. I considered it as a design option when I wrote my implementation (and voiced that possibility). I guess I hadn't fully decided, just put the decision off for later.

But let's look at the code you're afraid of:
```js
for (const f of fork(inputIterable)) {
  // ...
}
```
I don't think this code is really any cause for worry. First, the code is an infinite loop. If there's a break statement then we get to do our cleanup. If there's no break statement, then literally the whole rest of the program must be about this iterable, in which case the cache hardly seems like a leak or waste of memory.

I'd argue that you're also confusing our design intentions a bit. We put in the ability to discard unused cache items as an optimization for a specific use case, when the forks were ultimately going to be reintegrated item by item with `zip`, `merge`, or `collate`. The other use cause, however, where the user consumes some or all of each fork in sequence, is also perfectly valid and probably common. We don't need to fear that our optimization won't apply there, we just know that it wasn't meant to, because with that access pattern it doesn't really matter if we can or can't make more forks. If we can't make more forks, there must be some number `n` of forks that we made. With sequential access the cache won't be cleaned up at all until we're already iterating over the `n`th fork anyway, which is not much of an optimization.

**Comment by conartist6** [comment]
> The only valid reason to use iterables is that we are dealing with **a lot** of data. Possibly more than you can fit in your process.

I disagree again.

That isn't true either, not by a long shot. ES6 `Map` demonstrates another common one: Exposing data that you store without giving consumers the possibility of modifying it.

Immutable.js shows another: providing a compatible API that hides the complex internal structure that data really has.

And then there's working with data of more modest size when there are many intermediate transforms each of which would have to allocate memory for themselves, another good use case.

**Comment by sithmel** [comment]
> The other use cause, however, where the user consumes some or all of each fork in sequence, is also perfectly valid and probably common.

If I had to carry on operations in sequence I would rather use an array. Fork would be sub-optimal

> ES6 `Map` demonstrates another common one

My point was that iterables can used for representing infinite sequences, and for this reason I would avoid adding to iter-tools (that is a general purpose library) a function that completely disregard this use case.

> We don't need to fear that our optimization won't apply there, we just know that it wasn't meant to, because with that access pattern it doesn't really matter if we can or can't make more forks.

My concern is that the user won't understand that there is this trade-off. But perhaps writing a clear warning in the documentation will be enough.

**Comment by conartist6** [comment]
Yeah we should certainly express in documentation that there is reason to consider using `Array.from()` or `toArray` instead of fork, and that you need to be quite careful it you're using a fork around infinite iterables
